airflow/providers/databricks/plugins/databricks_workflow.py

```diff
@@ -0,0 +1,440 @@
+from airflow.utils.airflow_flask_app import AirflowApp
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.session import NEW_SESSION, provide_session
+from airflow.utils.state import TaskInstanceState
+from airflow.utils.task_group import TaskGroup
+from airflow.www.views import AirflowBaseView
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm.session import Session
+
+
+REPAIR_WAIT_ATTEMPTS = os.getenv("DATABRICKS_REPAIR_WAIT_ATTEMPTS", 20)
+REPAIR_WAIT_DELAY = os.getenv("DATABRICKS_REPAIR_WAIT_DELAY", 0.5)
+
+airflow_app = cast(AirflowApp, current_app)
+
+
+def _get_databricks_task_id(task: BaseOperator) -> str:
+    """Get the databricks task ID using dag_id and task_id. removes illegal characters.
+
+    :param task: The task to get the databricks task ID for.
+    :return: The databricks task ID.
+    """
+    return f"{task.dag_id}__" + task.task_id.replace(".", "__")
+
+
+def get_databricks_task_ids(
+    group_id: str, task_map: dict[str, BaseOperator], log: logging.Logger
+) -> list[str]:
+    """
+    Return a list of all Databricks task IDs for a dictionary of Airflow tasks.
+
+    :param group_id: The task group ID.
+    :param task_map: A dictionary mapping task IDs to BaseOperator instances.
+    :param log: The logger to use for logging.
+    :return: A list of Databricks task IDs for the given task group.
+    """
+    task_ids = []
+    log.debug("Getting databricks task ids for group %s", group_id)
+    for task_id, task in task_map.items():
+        if task_id == f"{group_id}.launch":
+            continue
+        databricks_task_id = _get_databricks_task_id(task)
+        log.debug("databricks task id for task %s is %s", task_id, databricks_task_id)
+        task_ids.append(databricks_task_id)
+    return task_ids
+
+
+@provide_session
+def _get_dagrun(dag: DAG, run_id, session=None) -> DagRun:
+    """
+    Retrieve the DagRun object associated with the specified DAG and run_id.
+
+    :param dag: The DAG object associated with the DagRun to retrieve.
+    :param run_id: The run_id associated with the DagRun to retrieve.
+    :param session: The SQLAlchemy session to use for the query. If None, uses the default session.
+    :return: The DagRun object associated with the specified DAG and run_id.
+    """
+    return session.query(DagRun).filter(DagRun.dag_id == dag.dag_id, DagRun.run_id == run_id).first()
+
+
+@provide_session
+def _clear_task_instances(dag_id: str, run_id: str, task_ids: list[str], log: logging.Logger, session=None):
+    dag = airflow_app.dag_bag.get_dag(dag_id)
+    log.debug("task_ids to clear", str(task_ids))
+    dr: DagRun = _get_dagrun(dag, run_id, session=session)
+    tis_to_clear = [ti for ti in dr.get_task_instances() if _get_databricks_task_id(ti) in task_ids]
+    clear_task_instances(tis_to_clear, session)
+
+
+def _repair_task(
+    databricks_conn_id: str,
+    databricks_run_id: int,
+    tasks_to_repair: list[str],
+    log: logging.Logger,
```

Comment:
not sure whether we sure rename the argument as `logger` instead

Reply:
done so